crates/db_views/src/person_content_combined_view.rs

```diff
@@ -0,0 +1,430 @@
+use crate::{
+  structs::{
+    CommentView,
+    LocalUserView,
+    PersonContentCombinedPaginationCursor,
+    PersonContentCombinedView,
+    PersonContentViewInternal,
+    PostView,
+  },
+  InternalToCombinedView,
+};
+use diesel::{
+  result::Error,
+  BoolExpressionMethods,
+  ExpressionMethods,
+  JoinOnDsl,
+  NullableExpressionMethods,
+  QueryDsl,
+  SelectableHelper,
+};
+use diesel_async::RunQueryDsl;
+use i_love_jesus::PaginatedQueryBuilder;
+use lemmy_db_schema::{
+  aliases::creator_community_actions,
+  newtypes::PersonId,
+  schema::{
+    comment,
+    comment_actions,
+    comment_aggregates,
+    community,
+    community_actions,
+    image_details,
+    local_user,
+    person,
+    person_actions,
+    person_content_combined,
+    post,
+    post_actions,
+    post_aggregates,
+  },
+  source::{
+    combined::person_content::{person_content_combined_keys as key, PersonContentCombined},
+    community::CommunityFollower,
+  },
+  utils::{actions, actions_alias, functions::coalesce, get_conn, DbPool},
+};
+use lemmy_utils::error::LemmyResult;
+
+impl PersonContentCombinedPaginationCursor {
+  // get cursor for page that starts immediately after the given post
+  pub fn after_post(view: &PersonContentCombinedView) -> PersonContentCombinedPaginationCursor {
+    let (prefix, id) = match view {
+      PersonContentCombinedView::Comment(v) => ('C', v.comment.id.0),
+      PersonContentCombinedView::Post(v) => ('P', v.post.id.0),
+    };
+    // hex encoding to prevent ossification
+    PersonContentCombinedPaginationCursor(format!("{prefix}{id:x}"))
+  }
+
+  pub async fn read(&self, pool: &mut DbPool<'_>) -> Result<PaginationCursorData, Error> {
+    let err_msg = || Error::QueryBuilderError("Could not parse pagination token".into());
+    let mut query = person_content_combined::table
+      .select(PersonContentCombined::as_select())
+      .into_boxed();
+    let (prefix, id_str) = self.0.split_at_checked(1).ok_or_else(err_msg)?;
+    let id = i32::from_str_radix(id_str, 16).map_err(|_err| err_msg())?;
+    query = match prefix {
+      "C" => query.filter(person_content_combined::comment_id.eq(id)),
+      "P" => query.filter(person_content_combined::post_id.eq(id)),
+      _ => return Err(err_msg()),
+    };
+    let token = query.first(&mut get_conn(pool).await?).await?;
+
+    Ok(PaginationCursorData(token))
+  }
+}
```

Comment:
You should be able to move this logic into a helper function which can be used by all the different types.

Reply:
Unfortunately I can't do that because each one of these has its own mapping to and from. But I'll create an issue to make a trait out of it at least.